Firestore/core/src/firebase/firestore/util/type_traits.h

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_TYPE_TRAITS_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_TYPE_TRAITS_H_
+
+#if __OBJC__
+#import <objc/objc.h>  // for id
+#endif
+
+#include <type_traits>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+#if __OBJC__
+
+/**
+ * A type trait that identifies whether or not the given type is an Objective-C
+ * class.
+ *
+ * is_objective_c_class<NSObject>::value == true
+ * is_objective_c_class<id>::value == true
+ *
+ * is_objective_c_class<int>::value == false
+ * is_objective_c_class<std::string>::value == false
+ */
+template <typename T>
+struct is_objective_c_class {
+ private:
+  using yes_type = char (&)[10];
+  using no_type = char (&)[1];
+
+  /**
+   * A non-existent function declared to produce a pointer to type T (which is
+   * consistent with the way Objective-C objects are referenced).
+   *
+   * Note that there is no definition for this function but that's okay because
+   * we only need it to reason about the function's type at compile type.
+   */
+  static T* Instance();
+
+  static yes_type Choose(id value);
+  static no_type Choose(...);
+
+ public:
+  using value_type = bool;
+
+  enum { value = sizeof(Choose(Instance())) == sizeof(yes_type) };
+
+  constexpr operator bool() const {
+    return value;
+  }
+
+  constexpr bool operator()() const {
+    return value;
+  }
+};
+
+template <>
+struct is_objective_c_class<id> : public std::true_type {};
```

Comment:
Why is this necessary in addition to `Choose(id)` overload?

Reply:
I added this because I wanted to be able to consider `id` to be an Objective-C type, but it didn't work naturally because even though `id` is a pointer type it does not match `T*` during overload resolution. 

Consider `T = NSString`:
  * `Instance()` returns `NSString*`
  * the compiler picks `Choose(id)` since Objective-C pointers are convertible to `id`.

Now consider `T = id`:
  * `Instance()` returns `id*` (a double-pointer)
  * the compiler picks `Choose(...)` since pointers to Objective-C pointers are not convertible to `id`.

However, your other feedback has prompted me to rethink this slightly: I generally don't care that the pointed-to thing is an Objective-C class, since instances are always referred to by pointer. What I actually care about is that the type of a given variable is a pointer to an Objective-C object.

Once examined through this lens, the trait that matters is `is_objective_c_pointer` and with a slight change in implementation this works naturally for `id` but no longer works correctly for `void`.

In any case, by naturally folding `id` into this pattern `id` no longer needs to be special cased in `StringFormat`'s constructors and so the `static_cast<id>` is no longer required for `id<Protocol>` references.

Thanks for pushing on this. I think the result is a great improvement.

Reply:
Thanks!

> the static_cast<id> is no longer required for id<Protocol> references.

Yes, this seems like a good feature.